### PR TITLE
Core: Add estimateRowCount for Files and Entries Metadata Tables

### PR DESCRIPTION
**Pull request by szehon-ho**
Spark queries of Metadata Tables will often try to check row count estimates repeatedly during planning phase, like picking a join strategy.

By default, this involves an extra call to read the manifest file header, as the default logic on the BaseContentScanTask is as follows:

```
  static long estimateRowsCount(long length, ContentFile<?> file) {
    long[] splitOffsets = splitOffsets(file);
    long splitOffset = splitOffsets != null ? splitOffsets[0] : 0L;
    double scannedFileFraction = ((double) length) / (file.fileSizeInBytes() - splitOffset);
    return (long) (scannedFileFraction * file.recordCount());
  }
```

However, we actually have this information in the metadata and can just return it directly without reading the file header.

This can save many calls on the Spark driver for planning (one call per manifest file, so potentially many thousands, in serial manner).  

This is useful as many Spark maintenance procedures involve querying these tables ( expireSnapshots, removeOrphanFiles)

### CODE REVIEW

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 1]
```diff
@@ -1753,4 +1753,35 @@ public void testPositionDeletesManyColumns() {
...
+    preparePartitionedTable(true);
+
+    assertEstimatedSize(new DataFilesTable(table));
+    assertEstimatedSize(new AllDataFilesTable(table));
+    assertEstimatedSize(new AllFilesTable(table));
+
+    if (formatVersion != 2) {
+      assertEstimatedSize(new DeleteFilesTable(table));
+      assertEstimatedSize(new AllDeleteFilesTable(table));
+    }
```
Comment (amogh-jahagirdar) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:1768]: Maybe misunderstanding something but shouldn't this be `if (formatVersion == 2)` for testing the delete files metadata tables?

Comment (szehon-ho) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:1768]: You are right, typo, i think there are no tasks so assert passed.  Let me rewrite it.

**Comment by szehon-ho** [comment]
cc @amogh-jahagirdar @RussellSpitzer does it make sense?  Thanks

**Review by amogh-jahagirdar — APPROVED** [review]
@szehon-ho This was a great find, this improvement makes sense to me! I'll wait in case @RussellSpitzer has any comments

File: core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java [hunk 2]
```diff
@@ -1753,4 +1753,36 @@ public void testPositionDeletesManyColumns() {
...
+    assertEstimatedSize(new AllEntriesTable(table));
+  }
+
+  private void assertEstimatedSize(Table metadataTable) throws Exception {
+    TableScan scan = metadataTable.newScan();
+
+    try (CloseableIterable<FileScanTask> tasks = scan.planFiles()) {
+      List<FileScanTask> taskList = Lists.newArrayList(tasks);
+      assertThat(taskList.size()).isGreaterThan(0);
+      taskList.forEach(task -> assertThat(task.estimatedRowsCount()).isEqualTo(4));
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableScans.java:1785]: Makes sense, ill make the change

**Review by RussellSpitzer — APPROVED** [review]
Looks good to me! Great find

**Comment by lurnagao-dahua** [comment]
Hi,May I ask why we are not using manifest.addedRowsCount() instead of manifest.addedFilesCount()?
deleted and existing is the same?

**Comment by szehon-ho** [comment]
This is for a metadata table and not data table.  Hence for manifest-read task, each manifest's row count is the number of files it references (added + deleted + existing)

**Comment by lurnagao-dahua** [comment]
> This is for a metadata table and not data table. Hence for manifest-read task, each manifest's row count is the number of files it references (added + deleted + existing)

Got it, thank you for your answer

**Pull request merged**